lib/ABI/FunctionType.cpp

```diff
@@ -724,50 +719,36 @@ class ConversionHelper {
 
   static DistributedArguments
   distributeNonPositionBasedArguments(const ArgumentContainer &Arguments,
-                                      std::size_t SkippedRegisters = 0) {
+                                      std::size_t SkippedRegisters) {
     DistributedArguments Result;
-    size_t UsedGeneralPurposeRegisterCount = SkippedRegisters;
-    size_t UsedVectorRegisterCount = 0;
+    size_t UsedGeneralPurposeRegisterCounter = SkippedRegisters;
+    size_t UsedVectorRegisterCounter = 0;
 
+    const abi::Definition &D = abi::predefined::get(ABI);
     for (const model::Argument &Argument : Arguments) {
       auto MaybeSize = Argument.Type.size();
       revng_assert(MaybeSize.has_value());
 
-      constexpr bool CanSplit = AT::ArgumentsCanBeSplitBetweenRegistersAndStack;
+      const bool CanSplit = D.ArgumentsCanBeSplitBetweenRegistersAndStack;
       if (Argument.Type.isFloat()) {
-        // The conventional non-position based approach is not applicable for
-        // vector registers since it's rare for multiple registers to be used
-        // to pass a single argument.
-        //
-        // For now, just return the next free register.
-        //
-        // TODO: handle this properly.
-
-        static constexpr auto &Registers = AT::VectorArgumentRegisters;
-        if (UsedVectorRegisterCount < Registers.size()) {
-          // There is a free register to put the argument in.
-          if (Result.size() <= Argument.Index)
-            Result.resize(Argument.Index + 1);
-
-          auto Register = Registers[UsedVectorRegisterCount];
-          Result[Argument.Index].Registers.emplace_back(Register);
-          Result[Argument.Index].Size = *MaybeSize;
-          Result[Argument.Index].SizeOnStack = 0;
-
-          UsedVectorRegisterCount++;
-        } else {
-          // There are no more free registers left,
-          // pass the argument on the stack.
-          if (Result.size() <= Argument.Index)
-            Result.resize(Argument.Index + 1);
-          Result[Argument.Index].Size = *MaybeSize;
-          Result[Argument.Index].SizeOnStack = paddedSizeOnStack(*MaybeSize);
-        }
+        const auto &Registers = D.VectorArgumentRegisters;
+        size_t &Counter = UsedVectorRegisterCounter;
+        const size_t Limit = 1;
+
+        auto [Distributed, NextIndex] = considerRegisters(*MaybeSize,
+                                                          Limit,
+                                                          Counter,
+                                                          Registers,
+                                                          CanSplit);
+        if (Result.size() <= Argument.Index)
+          Result.resize(Argument.Index + 1);
+        Result[Argument.Index] = Distributed;
+        Counter = NextIndex;
```

Comment:
I assume this commit and 4 next commits do not introduce semantic change. Is this the case?
Any reason not to squash them altogether?

Reply:
As long as they compile, I'm fine.

Reply:
They all compile, I make sure of it when going through the rebase yesterday.

Reply:
I think I'll close this then. Feel free to reopen if you feel like some of them should be merged.